pkg/resource_builders/envoyconfig/listeners.go

```diff
@@ -0,0 +1,260 @@
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/durationpb"
+	"google.golang.org/protobuf/types/known/structpb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+func ListenerHTTP_v1(desc envoyDynamicConfigDescriptor) (envoy.Resource, error) {
+	opts := desc.(*saasv1alpha1.ListenerHttp)
+
+	listener := &envoy_config_listener_v3.Listener{
+		Name:            desc.GetName(),
+		Address:         Address_v1("0.0.0.0", opts.Port),
+		ListenerFilters: ListenerFilters_v1(opts.CertificateSecretName != nil),
+		FilterChains: []*envoy_config_listener_v3.FilterChain{{
+			Filters: []*envoy_config_listener_v3.Filter{{
+				Name: "envoy.filters.network.http_connection_manager",
+				ConfigType: &envoy_config_listener_v3.Filter_TypedConfig{
+					TypedConfig: func() *anypb.Any {
+						any, err := anypb.New(
+							&http_connection_manager_v3.HttpConnectionManager{
+								AccessLog: AccessLogConfig_v1(desc.GetName(), opts.CertificateSecretName != nil),
+								CommonHttpProtocolOptions: func() *envoy_config_core_v3.HttpProtocolOptions {
+									po := &envoy_config_core_v3.HttpProtocolOptions{
+										IdleTimeout: durationpb.New(3600 * time.Second),
+									}
+									if opts.MaxConnectionDuration != nil {
+										po.MaxConnectionDuration = durationpb.New(opts.MaxConnectionDuration.Duration)
+									}
+									if opts.AllowHeadersWithUnderscores != nil && *opts.AllowHeadersWithUnderscores {
+										po.HeadersWithUnderscoresAction = envoy_config_core_v3.HttpProtocolOptions_ALLOW
+									} else {
+										po.HeadersWithUnderscoresAction = envoy_config_core_v3.HttpProtocolOptions_REJECT_REQUEST
+									}
+									return po
+								}(),
+
+								HttpFilters: HttpFilters_v1(opts.RateLimitOptions),
+								HttpProtocolOptions: func() *envoy_config_core_v3.Http1ProtocolOptions {
+									if opts.DefaultHostForHttp10 != nil {
+										return &envoy_config_core_v3.Http1ProtocolOptions{
+											AcceptHttp_10:         true,
+											DefaultHostForHttp_10: *opts.DefaultHostForHttp10,
+										}
+									}
+									return &envoy_config_core_v3.Http1ProtocolOptions{}
+								}(),
+								Http2ProtocolOptions: &envoy_config_core_v3.Http2ProtocolOptions{
+									MaxConcurrentStreams:        wrapperspb.UInt32(100),
+									InitialStreamWindowSize:     wrapperspb.UInt32(65536),   // 64 KiB
+									InitialConnectionWindowSize: wrapperspb.UInt32(1048576), // 1 MiB
+								},
+								RequestTimeout:    durationpb.New(300 * time.Second),
+								RouteSpecifier:    RouteConfigFromAds_v1(opts.RouteConfigName),
+								StatPrefix:        desc.GetName(),
+								StreamIdleTimeout: durationpb.New(300 * time.Second),
+								UseRemoteAddress:  wrapperspb.Bool(true),
+							})
+						if err != nil {
+							panic(err)
+						}
+						return any
+					}(),
+				},
+			}},
+		}},
+		PerConnectionBufferLimitBytes: wrapperspb.UInt32(32768), // 32 KiB
```

Comment:
This value is modifying current default one of 1MB, so 32KB is 31 times lower.

How do we know this much lower value won't cause any side effect?

Reply:
It's the recommendation for configuring envoy as edge proxy. It's a safety measure to avoid somebody opening hundreds of connections with huge payloads that could cause an out of memory issue. At worst it has a performance penalty for huge requests that are bigger than 32kb. Anyway, 32kb of text is humongous.

Reply:
I have analysed kibana logs from last 15 minutes, checking for `envoy.bytes_received `and `envoy.bytes_sent` fields greater than 32000 bytes, and found:


- Bytes received  (apicast): https://es.mgmt.3sca.net/_dashboards/goto/7584f19670e6ec49a5183eff52f5e02e

- Bytes sent  (apicast/backend-listener): https://es.mgmt.3sca.net/_dashboards/goto/3dffc20160f110990217b102ef9e1a63

Reply:
The size of the buffer does not limit the size of the request/response, at least not in the case of an http listener. The documentation explains [how the listener/cluster buffers work](https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/flow_control). It basically limits how much raw data can be read per read() call. On the downside, when decreasing the buffer size, for big requests/responses, several read() calls will be required instead of one. On the upside, it grants protection against untrusted downstreams. Just imagine the memory consumption if I open 1k concurrent requests with 1Mb of size each... basically a DoS because the envoy container will go out of memory because it will have 1Mb per connection * 1000 conections = 1Gb.

I just did a quick test with the development echo-api to ensure it just works even lowering the buffer size. I sent the saas-operator csv file as the request payload (more than 200kb) and it worked just fine:

```
-rw-r--r--. 1 roi roi 238K dic 12 16:50 bundle/manifests/saas-operator.clusterserviceversion.yaml
❯ curl -sk https://echo-api.dev.3sca.net -X POST -H "Content-Type: application/json" -d @bundle/manifests/saas-operator.clusterserviceversion.yaml | jq .headers

{
  "HTTP_VERSION": "HTTP/1.1",
  "HTTP_HOST": "echo-api.dev.3sca.net",
  "HTTP_USER_AGENT": "curl/7.82.0",
  "HTTP_ACCEPT": "*/*",
  "CONTENT_TYPE": "application/json",
  "CONTENT_LENGTH": "237564",
  "HTTP_X_FORWARDED_FOR": "37.223.143.55",
  "HTTP_X_FORWARDED_PROTO": "https",
  "HTTP_X_ENVOY_EXTERNAL_ADDRESS": "37.223.143.55",
  "HTTP_X_REQUEST_ID": "5e6541ab-a7e9-44c8-afba-ca128c7f763f",
  "HTTP_X_ENVOY_EXPECTED_RQ_TIMEOUT_MS": "15000"
}
```

Reply:
nice, fair enough!